### DISCUSSION TITLE
Custom moves: UndoMove approach.

### DISCUSSION THREAD

**Discussion by AlexeyTrefilov** [TimefoldAI/timefold-solver#517] [General]
Hello.
I'm trying to implement quite complicate custom move.
And looks like I should create another move to implement createUndoMove.
And this is a question: why should I do that?
I mean, is it not enough to use something like undoMoveOnGenuineVariables? Why should it be the whole move?

**Comment by triceo** [comment]
Hello @AlexeyTrefilov,

had we been designing this API today, maybe we'd do it differently. But at the time when custom moves were designed many years ago, this seemed like the best approach, and so this is what we have. I don't think we'll be changing this; after all, custom moves are not even public API.

But we are discussing a whole new way of doing moves - simpler, faster, more powerful. If and when that shows up, it remains to be seen. For the time being, custom moves is what we have.

**Reply by yuetan1988** [reply]
Hey @triceo 

Do we have the powerful doing moves now?
If you have any example about custom move example, that will be nice. Because I think I need a checkValidMove to remove some moves.  

Thanks  a lot.

**Reply by triceo** [reply]
There has been some movement on this front recently. Starting with Solver 1.16.0, undo moves are not necessary. Custom moves themselves are still there though.

**Reply by yuetan1988** [reply]
Thanks for the info. I will check it
